server/errorstore/errors.go

```diff
@@ -134,74 +134,17 @@ func sha256b64(s string) string {
 }
 
 func hashError(err error) string {
-	// Ok so the hashing process is as follows:
-	//
-	// a) we want to hash the type and error message of the *root* error (the
-	// last unwrapped error) so that if by mistake the same error is sent to
-	// Handler.Handle in multiple places in the code, after being wrapped
-	// differently any number of times, it still hashes to the same value
-	// (because the root is the same). The type is not sufficient because some
-	// errors have the same type but variable parts (e.g.  a struct value that
-	// implements the error interface and the message contains a file name that
-	// caused the error and that is stored in a struct field).
-	//
-	// b) in addition that a), we also want to hash all locations in the stack
-	// trace, so that the same error type and message (say, sql.ErrNoRows or
-	// io.UnexpectedEOF) caused at two different places in the code are not
-	// considered the same error. To get that location, the error must be wrapped
-	// at some point by eris.Wrap (or must be a user-created error via eris.New).
-	// We cannot hash only the leaf frame in the stack trace as that would all be
-	// ctxerr.New or ctxerr.Wrap (i.e. whatever common helper function used to
-	// create the eris error).
-	//
-	// c) if we call eris.Unpack on an error that is not *directly* an "eris"
-	// error (i.e. an error value returned from eris.Wrap or eris.New), then
-	// eris.Unpack will not return any location information. So if for example
-	// the error was wrapped with the pkg/errors.Wrap or the stdlib's fmt.Errorf
-	// calls at some point, eris.Unpack will not give us any location info. To
-	// get around this, we look for an eris-created error in the wrapped chain,
-	// and only give up hashing the location if we can't find any.
-	//
-	// d) there is no easy way to identify an "eris" error (i.e. we cannot simply
-	// use errors.As(err, <some Eris error type>)) as eris does not export its
-	// error type, and it actually uses 2 different internal error types. To get
-	// around this, we look for an error that has the `StackFrames() []uintptr`
-	// method, as both of eris internal errors implement that (see
-	// https://github.com/rotisserie/eris/blob/v0.5.1/eris.go#L182).
-
-	var sf interface{ StackFrames() []uintptr }
-	if errors.As(err, &sf) {
-		err = sf.(error)
-	}
-
-	unpackedErr := eris.Unpack(err)
-
-	if unpackedErr.ErrExternal == nil &&
-		len(unpackedErr.ErrRoot.Stack) == 0 &&
-		len(unpackedErr.ErrChain) == 0 {
-		return sha256b64(unpackedErr.ErrRoot.Msg)
-	}
+	cause, stack := ctxerr.Summarize(err)
 
 	var sb strings.Builder
-	if unpackedErr.ErrExternal != nil {
-		root := eris.Cause(unpackedErr.ErrExternal)
-		fmt.Fprintf(&sb, "%T\n%s\n", root, root.Error())
-	}
+	fmt.Fprintf(&sb, "%T\n%s\n", cause, cause.Error())
+	fmt.Fprintf(&sb, strings.Join(stack, "\n"))
```

Comment:
This hashing (based on `Summarize`) would hash differently if the same error was wrapped differently on the way back up the call stack, causing duplicates (at least in the sense that the root error would be a duplicate).

E.g. (pseudo code, don't typecheck it :D ):

```
func fail() error {
    return ctxerr.New(ctx, "fail") // root error
}

func doSomething() error {
    if err := fail(); err != nil {
        if someCondition {
            return ctxerr.Wrap(ctx, err, "someCondition was true") // causes err to be seen as different
        }
        return err
    }
}
```

Now, you may argue that `someCondition` (on the way back up the stack) makes it a different error. I think if it failed at the same code path, we should play it safe and treat it as the same error, we could have unforeseen cases where the number of different wrapping explodes (e.g. something highly variable like the time gets logged in a call to Wrap) and causes redis to fill up.

Reply:
Good observation 🤔 

> we should play it safe and treat it as the same error, we could have unforeseen cases where the number of different wrapping explodes (e.g. something highly variable like the time gets logged in a call to Wrap) and causes redis to fill up.

Currently, this could only happen if the error message and/or the stack trace contains a variable value (let's call it `t`), right? Under that assumption, even if the two errors in your example hash to the same given the same `t` we would still overwhelm Redis if `t` changes rapidly.

So, I think the problem is two-fold:

1. We need to find a way to ignore `t`. Maybe relying only on the stack trace is enough? The trace should always be constant for the same error given the same build, right?
2. We need a solid definition of what constitutes "the same code path" for two errors (and documenting that for reference) from your example, a way to slice it is "any wrap that happens in the same function, hashes to the same error" vs the current implementation which is "any wrap hashes to a different error"

My first thought (I have to think more about this) is to do `1` and keep the current behavior, which limits the number of errors to O(1) (with the constant being how many `return err` we have in the code)

Reply:
(note: not saying we should do the same as them, only pointing as a reference)

seems like Sentry hashes primarly on the trace, and then uses other stuff (like the error message) if a trace is not available:

> A stack trace is used by default if it's available. So if two events have the same stack trace, then they will group into one issue

source https://docs.sentry.io/product/sentry-basics/grouping-and-fingerprints/

Reply:
Hmm sorry I missed that `Summarize` returns only the root error (but the stack traces of all wraps). So what I said about the wrapping message that adds variation to the hash doesn't hold. However, every conditional Wrap will still add stack traces (and thus, change the hash), meaning that any conditional Wrap on the return up the stack could generate unforeseen duplicates.

That's why the previous implementation only captured (or at least intended to capture) the stack trace of the root error. I don't think other stack traces are very interesting - anything after the root error is returned is just going back up the stack (except perhaps for some temporary new calls back down while on the way up, but that's very unlikely, once an error is returned, all callers tend to return immediately).

So in addition to adding noise and size to the errors we store, capturing stack traces on each call to Wrap also changes the hash, which may cause duplicates when the wrapping is done based on some condition (e.g. sometimes we wrap, sometimes we don't).

Reply:
gotcha, thank you! I think I'm okay hashing only the topmost stack trace. I pointed my only concern to this [here](https://github.com/fleetdm/fleet/pull/5733#discussion_r874951361) but I'm not super set on it.

having said that, I think you raised a very good point we weren't considering so far: what if by mistake we put a variable that changes on each invocation in the error message? Feels like we should hash by trace only and use messages only if we don't have traces.

Reply:
I think that's a corner case that would be quite weird if the error message changed on every call, I wouldn't worry about that scenario.

That being said, we should always have a stack trace, no? The way I imagined this is that any call to `New/Errorf/Wrap/Wrapf/etc.` checks if we have the common metadata (with your version of the implementation, it could check if we have any `*FleetError` in the chain), and if not creates or wraps the error with one that has the metadata and stack trace. I guess as a final verification step, the call to `Handle` (or `Store`) could also check that and capture a (mostly useless, but still) stack trace?